contracts/sol6/reserves/KyberFprReserveV2.sol

```diff
@@ -0,0 +1,357 @@
+        IERC20 indexed destToken,
+        uint256 destAmount,
+        address payable destAddress
+    );
+    event TradeEnabled(bool enable);
+    event MaxGasPriceUpdated(uint128 newMaxGasPrice);
+    event WithdrawAddressApproved(IERC20 indexed token, address indexed addr, bool approve);
+    event NewTokenWallet(IERC20 indexed token, address indexed wallet);
+    event WithdrawFunds(IERC20 indexed token, uint256 amount, address indexed destination);
+    event SetContractAddresses(
+        address indexed network,
+        IConversionRates indexed rate,
+        IWeth weth,
+        IKyberSanity sanity
+    );
+
+    constructor(
+        address _kyberNetwork,
+        IConversionRates _ratesContract,
+        IWeth _weth,
+        uint128 _maxGasPriceWei,
+        address _admin
+    ) public Withdrawable3(_admin) {
+        require(_kyberNetwork != address(0), "kyberNetwork 0");
+        require(_ratesContract != IConversionRates(0), "ratesContract 0");
+        require(_weth != IWeth(0), "weth 0");
+        kyberNetwork = _kyberNetwork;
+        conversionRatesContract = _ratesContract;
+        weth = _weth;
+        configData = ConfigData({tradeEnabled: true, maxGasPriceWei: _maxGasPriceWei});
+    }
+
+    receive() external payable {
+        emit DepositToken(ETH_TOKEN_ADDRESS, msg.value);
+    }
+
+    function trade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool validate
+    ) external override payable returns (bool) {
+        validate;
+        require(configData.tradeEnabled, "trade not enable");
+        require(msg.sender == kyberNetwork, "wrong sender");
+        require(tx.gasprice <= uint256(configData.maxGasPriceWei), "gas price too high");
+
+        doTrade(srcToken, srcAmount, destToken, destAddress, conversionRate);
+
+        return true;
+    }
+
+    function enableTrade() external onlyAdmin {
+        configData.tradeEnabled = true;
+        emit TradeEnabled(true);
+    }
+
+    function disableTrade() external onlyAlerter {
+        configData.tradeEnabled = false;
+        emit TradeEnabled(false);
+    }
+
+    function setMaxGasPrice(uint128 newMaxGasPrice) external onlyOperator {
+        configData.maxGasPriceWei = newMaxGasPrice;
+        emit MaxGasPriceUpdated(newMaxGasPrice);
+    }
+
+    function approveWithdrawAddress(
+        IERC20 token,
+        address addr,
+        bool approve
+    ) external onlyAdmin {
+        approvedWithdrawAddresses[keccak256(abi.encodePacked(address(token), addr))] = approve;
+        setDecimals(token);
+        emit WithdrawAddressApproved(token, addr, approve);
+    }
+
+    /// @dev allow set tokenWallet[token] back to 0x0 address
+    function setTokenWallet(IERC20 token, address wallet) external onlyAdmin {
+        tokenWallet[address(token)] = wallet;
+        setDecimals(token);
+        emit NewTokenWallet(token, wallet);
```

Comment:
should we add a special event for eth?
also wondering should it warn if reserve has eth balance?

Reply:
I added the dev comment, reserve manager that uses this reserve should have knowledge of how it works

Reply:
OK
good enough.